File: python/src/threads/postgres/dsn.py
"""The connection string holds a password (DATABASE_URL): no error may carry it. A malformed one is
invalid_config naming nothing of it; the driver's connect errors are scrubbed of the string and
its password, and never chained to psycopg's own."""

import re
from itertools import pairwise
from urllib.parse import unquote, urlsplit

_PASSWORD = re.compile(r"password\s*=\s*('(?:\\.|[^'])*'|\S+)")


def scrub(text: str, dsn: str) -> str:
    """`text` without `dsn` or any password in it."""
    for secret in sorted(_secrets(dsn), key=len, reverse=True):
        text = text.replace(secret, "[redacted]")
    return text


def _secrets(dsn: str) -> set[str]:
    found = {dsn}
    for match in _PASSWORD.finditer(dsn):
        value = match.group(1)
        found |= {value, value.strip("'")}
    words = dsn.split()
    for before, after in pairwise(words):
        if before == "password":  # a malformed `password value`
            found.add(after)
    try:
        password = urlsplit(dsn).password
    except ValueError:
        password = None
    if password:
        found |= {password, unquote(password)}
    return {s for s in found if s}

File: typescript/packages/postgres/src/dsn.ts
import { ConfigError } from "@threads/core/adapter";

// The connection URL holds a password (DATABASE_URL): no error may carry it. A string that isn't
// a postgres:// URL is refused without echoing it; the driver's own errors are scrubbed of the
// URL and its password before they leave the store.

/** The URL, parsed; invalid_config, naming nothing of it, when it isn't a postgres:// URL. */
export function postgresUrl(dsn: string): URL {
  const url = URL.canParse(dsn) ? new URL(dsn) : undefined;
  if (
    url === undefined ||
    (url.protocol !== "postgres:" && url.protocol !== "postgresql:")
  )
    throw new ConfigError(
      "invalid_config",
      "postgres() needs a postgres:// URL",
    );
  return url;
}

/** Removes `dsn` and its password from `error`'s message and stack, and every cause's. */
export function scrubbed(error: unknown, dsn: string): unknown {
  const secrets = secretsOf(dsn);
  for (let at = error; at instanceof Error; at = at.cause) {
    at.message = scrub(at.message, secrets);
    if (at.stack !== undefined) at.stack = scrub(at.stack, secrets);
  }
  return error;
}

function secretsOf(dsn: string): readonly string[] {
  const password = URL.canParse(dsn) ? new URL(dsn).password : "";
  return [dsn, password, safeDecode(password)]
    .filter((s) => s.length > 0)
    .toSorted((a, b) => b.length - a.length);
}

function safeDecode(text: string): string {
  try {
    return decodeURIComponent(text);
  } catch {
    return text;
  }
}

function scrub(text: string, secrets: readonly string[]): string {
  return secrets.reduce((out, s) => out.replaceAll(s, "[redacted]"), text);
}
